File: android/app/src/main/kotlin/com/cruisemesh/app/mesh/LanHealthTracker.kt
package com.cruisemesh.app.mesh

import uniffi.cruisemesh_core.CoreLanHealthAction
import uniffi.cruisemesh_core.CoreLanHealthTracker

internal class LanHealthTracker(timeoutMs: Long = 20_000, maxConsecutiveTimeouts: Int = 3) {
    sealed interface Decision {
        data class Send(val nonce: ULong) : Decision
        data object Wait : Decision
        data object Close : Decision
    }

    private val core = CoreLanHealthTracker(timeoutMs, maxConsecutiveTimeouts.toUInt())

    fun next(address: String, nowMs: Long, nonce: ULong): Decision {
        val decision = core.next(address, nowMs, nonce)
        return when (decision.action) {
            CoreLanHealthAction.SEND -> Decision.Send(requireNotNull(decision.nonce))
            CoreLanHealthAction.WAIT -> Decision.Wait
            CoreLanHealthAction.CLOSE -> Decision.Close
        }
    }

    fun response(address: String, nonce: ULong, nowMs: Long): Long? = core.response(address, nonce, nowMs)
    fun remove(address: String) = core.remove(address)
    fun clear() = core.clear()
}

File: ios/CruiseMesh/Mesh/LanHealthTracker.swift
import Foundation

final class LanHealthTracker {
    enum Decision: Equatable { case send(UInt64), wait, close }
    private let core: CoreLanHealthTracker

    init(timeoutMs: Int64 = 20_000, maxConsecutiveTimeouts: Int = 3) {
        core = CoreLanHealthTracker(timeoutMs: timeoutMs, maxTimeouts: UInt32(maxConsecutiveTimeouts))
    }

    func next(address: String, nowMs: Int64, nonce: UInt64) -> Decision {
        let decision = core.next(address: address, nowMs: nowMs, nonce: nonce)
        switch decision.action {
        case .send: return .send(decision.nonce!)
        case .wait: return .wait
        case .close: return .close
        }
    }

    func response(address: String, nonce: UInt64, nowMs: Int64) -> Int64? {
        core.response(address: address, nonce: nonce, nowMs: nowMs)
    }
    func remove(address: String) { core.remove(address: address) }
    func clear() { core.clear() }
}
